src/Compilers/CSharp/Portable/Syntax/CSharpSyntaxTree.LazySyntaxTree.cs

```diff
@@ -0,0 +1,188 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Text;
+using System.Threading;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    public partial class CSharpSyntaxTree
+    {
+        private class LazySyntaxTree : CSharpSyntaxTree
+        {
+            private readonly CSharpParseOptions _options;
+            private readonly string _path;
+            private readonly SourceText _text;
+            private readonly ImmutableDictionary<string, ReportDiagnostic> _diagnosticOptions;
+            private CSharpSyntaxNode? _lazyRoot;
+
+            internal LazySyntaxTree(
+                SourceText text,
+                string path,
+                CSharpParseOptions options,
+                ImmutableDictionary<string, ReportDiagnostic>? diagnosticOptions)
+            {
+                Debug.Assert(options != null);
+
+                _text = text;
+                _options = options;
+                _path = path ?? string.Empty;
+                _diagnosticOptions = diagnosticOptions ?? EmptyDiagnosticOptions;
+            }
+
+            public override string FilePath
+            {
+                get { return _path; }
+            }
+
+            public override SourceText GetText(CancellationToken cancellationToken)
+            {
+                return _text;
+            }
+
+            public override bool TryGetText([NotNullWhen(true)] out SourceText? text)
+            {
+                text = _text;
+                return true;
+            }
+
+            public override Encoding? Encoding
+            {
+                get { return _text.Encoding; }
+            }
+
+            public override int Length
+            {
+                get { return _text.Length; }
+            }
+
+            public override CSharpSyntaxNode GetRoot(CancellationToken cancellationToken)
+            {
+                if (!TryGetRoot(out var root))
```

Comment:
I agree, this could be simpler with `if (_lazyRoot != null) { var root = ...; _ = Interlocked.CompareExchange(ref _lazyRoot, root, null); } return _lazyRoot;`

---
In reply to: [569688252](https://github.com/dotnet/roslyn/pull/50490#discussion_r569688252) [](ancestors = 569688252)